packages/lib/server/checkDurationLimits.ts

```diff
@@ -0,0 +1,95 @@
+import dayjs from "@calcom/dayjs";
+import prisma from "@calcom/prisma";
+
+import { HttpError } from "../http-error";
+import { parseDurationLimit } from "../isDurationLimits";
+
+export async function checkDurationLimits(durationLimits: any, eventStartDate: Date, eventId: number) {
+  const parsedDurationLimits = parseDurationLimit(durationLimits);
+  if (!parsedDurationLimits) {
+    return false;
+  }
+
+  const limitCalculations = Object.entries(parsedDurationLimits).map(
+    async ([key, limitingNumber]) =>
+      await checkDurationLimit({ key, limitingNumber, eventStartDate, eventId })
+  );
+
+  await Promise.all(limitCalculations).catch((error) => {
+    throw new HttpError({ message: error.message, statusCode: 401 });
+  });
+
+  return true;
+}
+
+export async function checkDurationLimit({
+  eventStartDate,
+  eventId,
+  key,
+  limitingNumber,
+  returnBusyTimes = false,
+}: {
+  eventStartDate: Date;
+  eventId: number;
+  key: string;
+  limitingNumber: number;
+  returnBusyTimes?: boolean;
+}) {
+  {
+    // Take PER_DAY and turn it into day and PER_WEEK into week etc.
+    const filter = key.split("_")[1].toLocaleLowerCase() as "day" | "week" | "month" | "year";
+    const startDate = dayjs(eventStartDate).startOf(filter).toDate();
+    const endDate = dayjs(startDate).endOf(filter).toDate();
+
+    const bookingTimes = await prisma.booking.findMany({
+      select: {
+        startTime: true,
+        endTime: true,
+      },
+      where: {
+        status: "ACCEPTED",
+        eventType: {
+          AND: [
+            {
+              id: eventId,
+              bookings: {
+                every: {
+                  startTime: {
+                    gte: startDate,
+                  },
+                  endTime: {
+                    lte: endDate,
+                  },
+                },
+              },
+            },
+          ],
+        },
+      },
+    });
+
+    const totalBookingDuration = sumDuration(bookingTimes);
+    if (totalBookingDuration >= limitingNumber) {
+      // This is used when getting availability
+      if (returnBusyTimes) {
+        return {
+          start: startDate,
+          end: endDate,
+        };
+      }
+
+      throw new HttpError({
+        message: `duration_limit_reached`,
+        statusCode: 403,
+      });
+    }
+  }
+}
+
+// This is a helper function to sum the duration of all bookings
+const sumDuration = (bookings: { startTime: Date; endTime: Date }[]) => {
+  return bookings.reduce((acc, booking) => {
+    const duration = dayjs(booking.endTime).diff(booking.startTime, "minute");
+    return acc + duration;
+  }, 0);
+};
```

Comment:
If this is a whole year of bookings it might take a bit to compute this for every day this is called ... 

@emrysal any ideas the best approach here? We might be best thinking about how we handle this as this.

Reply:
Relooking at the code i believe this is only called on creation on the event so isnt as bad as i thought. 

We instead do this https://github.com/calcom/cal.com/pull/7166/files#diff-3bcd8b09138a03622b300b58836dfed0598874aed41ff92f57c2b2700f30f4b5R303 in getUserAvaliblility  instead

Reply:
I managed to find a way to delegate this task to the data using a raw SQL query, here's a snippet of something that I'm working on right now.

```tsx
const getTotalBookingDuration = async (eventId: number, startDate: Date, endDate: Date) => {
  // Aggregates the total booking time for a given event in a given time period
  const [totalBookingTime] = (await prisma.$queryRaw`
    SELECT SUM(EXTRACT(EPOCH FROM ("endTime" - "startTime")) / 60) as "totalMinutes"
    FROM "Booking"
    WHERE "status" = 'accepted'
      AND "id" = ${eventId}
      AND "startTime" >= ${startDate}
      AND "endTime" <= ${endDate};
  `) as { totalMinutes: number }[];

  return totalBookingTime.totalMinutes;
};
```

Reply:
CC @zomars

Reply:
Just pushed some additional commits, these two particularly tries to address this:

https://github.com/calcom/cal.com/pull/7166/commits/9820711ebfa26ec70af08aea033382eced626cfc
https://github.com/calcom/cal.com/pull/7166/commits/1ad8b48152b607f18355fe84929737dba23bebcc

Would love to hear your thoughts